packages/@react-stately/grid/src/useGridState.ts

```diff
@@ -46,17 +46,67 @@ export function useGridState<T extends object, C extends GridCollection<T>>(prop
     setFocusedKey(key, child);
   };
 
+  let selectionManager = useMemo(() =>
+    new SelectionManager(collection, selectionState)
+    , [collection, selectionState]
+  );
+
   // Reset focused key if that item is deleted from the collection.
+  const cachedCollection = useRef(null);
   useEffect(() => {
     if (selectionState.focusedKey != null && !collection.getItem(selectionState.focusedKey)) {
-      selectionState.setFocusedKey(null);
+      const node = cachedCollection.current.getItem(selectionState.focusedKey);
+      const parentNode =
+        node.parentKey != null && (node.type === 'cell' || node.type === 'rowheader' || node.type === 'column') ?
+        cachedCollection.current.getItem(node.parentKey) :
+        node;
+      const cachedRows = cachedCollection.current.rows;
+      const rows = collection.rows;
+      const diff = cachedRows.length - rows.length;
+      let index = Math.min(
+        (
+          diff > 1 ?
+          Math.max(parentNode.index - diff + 1, 0) :
```

Comment:
So if you deleted a row and at the same time added 2+ more the index would change but otherwise it stays the same? What was the reason for this?

Reply:
Was added to try to address the concerns @LFDanLu raised with [the ListBox example](https://reactspectrum.blob.core.windows.net/reactspectrum/cda282acb88ee14256fc78ba8f8c772b5e6d16dc/storybook/index.html?path=/story/listbox--restore-focus-after-deleting-selected-items): https://github.com/adobe/react-spectrum/pull/3885#discussion_r1072940153.

When you remove multiple items, and depending on where focus is in relation to the items selected for removal, one perceives the focus jumping ahead significantly from where it was. This code intends to mitigate that perception. Do you think there is a better approach to this?

Reply:
> So if you deleted a row and at the same time added 2+ more the index would change but otherwise it stays the same? What was the reason for this?

Also, in this use case the `diff` would be less than 1, because the new row count would be greater than the cached row count, `const diff = cachedRows.length - rows.length;`, so we would just use `Math.min(parentNode.index,  rows.length - 1)`.

Reply:
Ok that makes sense. One weird case I see is if you select two items from the bottom up so focus is on the first one, focus moves to the previous item rather than the next one. For example, select Snake and Kangaroo, delete, and focus ends on Aardvark rather than Danni (which is now at the corresponding index as Snake). If you select the opposite direction so focus is on Snake, then focus moves to Danni instead (rather than Devon which is at the corresponding index).

Reply:
I kinda like that behavior since it is restoring focus to the closest item in the direction the user's focus was closest/trending to if that makes sense (albeit that isn't what the code is explicitly doing). There is another odd case that I noted [here](https://github.com/adobe/react-spectrum/pull/3885#pullrequestreview-1266653110), so I think it would be good for the team to test this vs the previous implementation for opinions/edge cases and maybe we can make a call on if we are fine with this or if we want to refine some more

Reply:
Mac Finder:
Select two, bottom first. Delete. Loses selection entirely.
Select two, top first. Delete. Selects item that replaced first selected index, not the focused one.
Select one. Delete. Selects the item that followed it.
Select two, top first, gap of n items between. Selects item that followed last focused+selected item.
Select two, bottom first, gap of n items between. Selects item that followed last focused+selected item.


Could improve on this by not losing selection in the first case and sending it to the next available item (checking more than one down in otherwords). The rest seems fine to use as the template for how to move?